lib/Conversion/TorchToMhlo/ReductionOp.cpp

```diff
@@ -0,0 +1,572 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// Also available under a BSD-style license. See LICENSE.
+//
+//===----------------------------------------------------------------------===//
+
+#include "torch-mlir/Conversion/TorchToMhlo/TorchToMhlo.h"
+
+#include "../PassDetail.h"
+#include "./MhloLegalizeUtils.h"
+#include "./PopulatePatterns.h"
+#include "mlir-hlo/Dialect/mhlo/IR/hlo_ops.h"
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "torch-mlir/Conversion/Utils/Utils.h"
+#include "torch-mlir/Dialect/Torch/IR/TorchDialect.h"
+#include "torch-mlir/Dialect/Torch/IR/TorchOps.h"
+#include "torch-mlir/Dialect/Torch/Utils/TorchUpstream.h"
+#include "torch-mlir/Dialect/Torch/Utils/Utils.h"
+#include "torch-mlir/Dialect/TorchConversion/IR/TorchConversionOps.h"
+
+using namespace mlir;
+using namespace mlir::torch;
+using namespace mlir::torch::Torch;
+
+static Value createInitialValueForReduceOp(Operation *op, Type elementTy,
+                                           PatternRewriter &rewriter) {
+  auto constType = RankedTensorType::get({}, elementTy);
+  if (isa<AtenSumOp, AtenSumDimIntListOp>(op)) {
+    if (elementTy.isa<mlir::FloatType>()) {
+      auto constAttr = DenseElementsAttr::get(
+          constType, {APFloat::getZero(
+                         elementTy.cast<mlir::FloatType>().getFloatSemantics(),
+                         /*negative=*/false)});
+      return rewriter.create<mhlo::ConstantOp>(op->getLoc(), constType,
+                                               constAttr);
+    } else if (elementTy.isa<mlir::IntegerType>() &&
+               elementTy.getIntOrFloatBitWidth() != 8) {
+      auto constAttr = DenseElementsAttr::get(
+          constType, {APInt::getZero(elementTy.getIntOrFloatBitWidth())});
+      return rewriter.create<mhlo::ConstantOp>(op->getLoc(), constType,
+                                               constAttr);
+    }
+  }
+
+  if (isa<AtenMaxOp, AtenMaxDimOp, AtenArgmaxOp>(op)) {
+    if (elementTy.isa<mlir::FloatType>()) {
+      auto constAttr = DenseElementsAttr::get(
+          constType, {APFloat::getLargest(
+                         elementTy.cast<mlir::FloatType>().getFloatSemantics(),
+                         /*negative=*/true)});
+      return rewriter.create<mhlo::ConstantOp>(op->getLoc(), constType,
+                                               constAttr);
+    } else if (elementTy.isa<mlir::IntegerType>() &&
+               elementTy.getIntOrFloatBitWidth() != 8) {
+      auto constAttr = DenseElementsAttr::get(
+          constType,
+          {APInt::getSignedMinValue(elementTy.getIntOrFloatBitWidth())});
+      return rewriter.create<mhlo::ConstantOp>(op->getLoc(), constType,
+                                               constAttr);
+    }
+  }
+
+  op->emitError("unimplemented lowering in "
```

Comment:
The return value of this function is never checked if null and seems only called where it would return a non-null Value, either it should be checked if null on return or llvm_unreachable used to assert rather than null dereference below.